posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+
+```tut:book:silent
+class UserRoutesMTL[F[_]: Sync](userAlgebra: UserAlgebra[F])(implicit ev: HttpErrorHandler[F, UserError]) extends Http4sDsl[F] {
+
+  import JsonCodecs._
+  import syntax._
+
+  val routes: HttpRoutes[F] = HttpRoutes.of[F] {
+
+    case GET -> Root / "users" / username =>
+      userAlgebra.find(username).flatMap {
+        case Some(user) => Ok(user)
+        case None => NotFound(username)
+      }.handleHttpErrorResponse
+
+    case req @ POST -> Root / "users" =>
+      req.as[User].flatMap { user =>
+        userAlgebra.save(user) *> Created(user.username)
+      }.handleHttpErrorResponse
+
+    case req @ PUT -> Root / "users" / username =>
+      req.as[UserUpdateAge].flatMap { userUpdate =>
+        userAlgebra.updateAge(username, userUpdate.age) *> Created(username)
+      }.handleHttpErrorResponse
+  }
+
+}
+```
+
+We are basically delegating the error handling (AKA mapping business errors to appropiate http responses) to a specific algebra by making use of the syntax we have previously introduced.
+
+We also need an implementation for this algebra in order to handle errors of type `UserError`.
+
+```tut:book:silent
+class UserHttpErrorHandler[F[_]](implicit M: MonadError[F, UserError]) extends HttpErrorHandler[F, UserError] with Http4sDsl[F] {
+  override def handle(fa: F[Response[F]]): F[Response[F]] =
+    fa.handleErrorWith {
+      case InvalidUserAge(age) => BadRequest(s"Invalid age $age")
+      case UserAlreadyExists(username) => Conflict(username)
+      case UserNotFound(username) => NotFound(username)
+    }
+}
+```
+
+If we forget to handle some errors the compiler will shout at us ***"match may not be exhaustive!"*** That's fantastic :)
+
+#### Wiring all the components
+
+And the last part will be the wiring of all these components where we need to include the `meow-mtl` import to figure out the derivation of the instances we need in order to make this work. It'll look something like this if using `cats.effect.IO`:
+
+```scala
+import cats.effect.IO
+import com.olegpy.meow.hierarchy._
+
+implicit val userHttpErrorHandler: HttpErrorHandler[IO, UserError] = new UserHttpErrorHandler[IO]
+
+UserInterpreter.create[IO].flatMap { UserAlgebra =>
+  routes = new UserRoutesMTL[IO](UserAlgebra)
+  IO.unit // pretend this is the rest of your program
+}
+```
+
+### Final thoughts
+
+This is such an exciting time to be writing pure functional programming in Scala! The Typelevel ecosystem is getting richer and more mature having an amazing set of libraries to solve business problems in an elegant and purely functional way.
+
+I hope you have enjoyed this post and please do let me know if you know of better ways to solve this problem in the comments!
+
+And last but not least I would like to thank all the friendly folks I hang out with in the `cats-effect`, `cats`, `fs2` and `http4s` Gitter channels for all the time and effort they put (*for free*) into making this community an amazing space.
```

Comment:
And many thanks to you for your contributions in all these spaces.  We :heart: good documenters.